File: ReactNativeApp/android/app/src/main/java/com/phases/instances/GamePhase.java
package com.phases.instances;

import com.components.basic.points.PointsBuilder;
import com.components.basic.points.PointsController;
import com.components.basic.points.PointsState;
import com.components.basic.points.PointsView;
import com.components.basic.streak.StreakBuilder;
import com.components.basic.streak.StreakController;
import com.components.basic.streak.StreakState;
import com.components.basic.streak.StreakView;
import com.components.basic.timer.TimerBuilder;
import com.components.basic.timer.TimerController;
import com.components.basic.timer.TimerState;
import com.components.basic.timer.TimerView;
import com.phases.base.Phase;
import com.phases.phasemanager.PhaseManager;
import com.ssvc.SSVC;

public class GamePhase extends Phase {

    protected SSVC<PointsState, PointsView, PointsController> points;
    protected SSVC<TimerState, TimerView, TimerController> timer;
    protected SSVC<StreakState, StreakView, StreakController> streak;


    @Override
    public void initialize() {
        this.prevPhase = PhaseManager.GamePhases.CALIBRATION_PHASE;
        this.nextPhase = PhaseManager.GamePhases.END_OF_EXERCISE_PHASE;
        points = PointsBuilder.buildPoints();
        timer = TimerBuilder.buildTimer();
        streak = StreakBuilder.build();
    }

    @Override
    public PhaseManager.GamePhases getNextPhase() {
        return this.nextPhase;
    }

    @Override
    public PhaseManager.GamePhases getPrevPhase() {
        return this.prevPhase;
    }

    @Override
    public void process() {
    }

    @Override
    public boolean isDone() {
        return !timer.state.running && timer.state.finished;
    }


    private void timerCleanUp() {
        timer.state.cleanUp();
        timer.controller.stopCountDownTimer();
        timer.stateView.cleanUp();
        timer.stateView.detachFromLayoutSV();
    }

    public void cleanup() {
        super.cleanup();
        timerCleanUp();
        points.stateView.cleanUp();
        streak.controller.cleanUp();
        streak.stateView.cleanUp();

    }

}

File: ReactNativeApp/ios/Modules/Phases/Instances/GamePhase.swift
//
//  GamePhase.swift
//  jogo
//
//  Created by Mohsin on 13/10/2021.
//

import Foundation

public class GamePhase: Phase {
  
  var points: SSVC<PointsState, SquaredPointsView, PointsController>?
  var timer: SSVC<TimerState, TimerView, TimerController>?
  var ball: BallSSVC<BallState, BallView, BallController>?
  var person: PersonSSVC<PersonState, PersonView, PersonController>?
  
  
  
  public override func process() {
    timer?.controller.process()
    super.process()
  }
  
  /// Provides the functionality of returning the completion of this phase as well as writing down
  /// the value of the complete elapsed time of the game phase.
  public override func isDone() -> Bool {
    let isDone: Bool = timer?.controller.timeFinished() ?? true
    if isDone {
      ExerciseStatsJsonManager.get().fillProperties(completionTime: timer?.controller.getCompletionTime() ?? 0.0)
      ExerciseStatsJsonManager.get().convertAndWriteToFile()
    }
    return isDone
  }
  
  private static var gamePhase: GamePhase?
  
  public static func getPhase() -> GamePhase? {
    return GamePhase.gamePhase
  }
  
  public override func initialize() {
    self.prevPhase = .CalibrationPhase
    self.nextPhase = .EndOfExercisePhase
    points = PointsBuilder.buildSquared()
    timer = TimerBuilder.build()
    person = PersonBuilder.build(state: GameContext.getContext()!.personState!)
    ball = BallBuilder.build(state: GameContext.getContext()!.ballState!)
    person?.changeView(drawType: .SKELETON)
    ball?.changeView(drawType: .CIRCLE)
    GamePhase.gamePhase = self
    super.initialize()
    GameContext.getContext()?.exerciseTime = Date().getMilliseconds()
    
  }
  
  public override func getDebugText() -> String {
    return "\n\n Points:- \n \(points?.getDebugText() ?? "") \n\n Timer:- \n \(timer?.getDebugText() ?? "") \n\n"
  }
  
  public override func cleanup() {
    super.cleanup()
    GameContext.getContext()?.currentScore = self.points?.controller.getPoints() ?? 0
    self.points?.cleanup()
    self.timer?.cleanup()
    self.ball?.cleanup(with: false)
    self.person?.cleanup(with: false)
    self.points = nil
    self.timer = nil
    self.ball = nil
    self.person = nil
  }
}
